// thor-devkit/dist/hdnode.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.HDNode = void 0;
const basex_1 = require("@vechain/ethers/utils/basex");
const HD = __importStar(require("@vechain/ethers/utils/hdnode"));
const crypto_1 = require("crypto");
const elliptic_1 = require("elliptic");
const address_1 = require("./address");
const buffer_1 = require("buffer");
// see https://github.com/satoshilabs/slips/blob/master/slip-0044.md
const VET_DERIVATION_PATH = `m/44'/818'/0'/0`;
const xpubPrefix = buffer_1.Buffer.from('0488b21e000000000000000000', 'hex');
const xprivPrefix = buffer_1.Buffer.from('0488ade4000000000000000000', 'hex');
const curve = new elliptic_1.ec('secp256k1');
var HDNode;
(function (HDNode) {
    /** create node from mnemonic words */
    function fromMnemonic(words, path = VET_DERIVATION_PATH) {
        // normalize words to lowercase
        const joinedWords = words.join(' ').toLowerCase();
        const node = HD.fromMnemonic(joinedWords).derivePath(path);
        return createHDNode(node);
    }
    HDNode.fromMnemonic = fromMnemonic;
    /**
     * create node from xpub
     * @param pub public key
     * @param chainCode chain code
     */
    function fromPublicKey(pub, chainCode) {
        const compressed = curve.keyFromPublic(pub).getPublic(true, 'array');
        const key = buffer_1.Buffer.concat([xpubPrefix, chainCode, buffer_1.Buffer.from(compressed)]);
        const checksum = sha256(sha256(key)).slice(0, 4);
        const node = HD.fromExtendedKey(basex_1.Base58.encode(buffer_1.Buffer.concat([key, checksum])));
        return createHDNode(node);
    }
    HDNode.fromPublicKey = fromPublicKey;
    /**
     * create node from xpriv
     * @param priv private key
     * @param chainCode chain code
     */
    function fromPrivateKey(priv, chainCode) {
        const key = buffer_1.Buffer.concat([xprivPrefix, chainCode, buffer_1.Buffer.from([0]), priv]);
        const checksum = sha256(sha256(key)).slice(0, 4);
        const node = HD.fromExtendedKey(basex_1.Base58.encode(buffer_1.Buffer.concat([key, checksum])));
        return createHDNode(node);
    }
    HDNode.fromPrivateKey = fromPrivateKey;
    function createHDNode(ethersNode) {
        const pub = buffer_1.Buffer.from(curve.keyFromPublic(ethersNode.publicKey.slice(2), 'hex').getPublic(false, 'array'));
        const priv = ethersNode.privateKey ? buffer_1.Buffer.from(ethersNode.privateKey.slice(2), 'hex') : null;
        const cc = buffer_1.Buffer.from(ethersNode.chainCode.slice(2), 'hex');
        const addr = address_1.address.fromPublicKey(pub);
        return {
            get publicKey() {
                return pub;
            },
            get privateKey() {
                return priv;
            },
            get chainCode() {
                return cc;
            },
            get address() {
                return addr;
            },
            derive(index) {
                return createHDNode(ethersNode.derivePath('' + index));
            }
        };
    }
    function sha256(data) {
        return crypto_1.createHash('sha256').update(data).digest();
    }
})(HDNode = exports.HDNode || (exports.HDNode = {}));
