// @vechain/connex/esm/driver.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
/* eslint-disable @typescript-eslint/no-unused-vars */
import { DriverNoVendor, SimpleNet } from '@vechain/connex-driver';
import { blake2b256 } from 'thor-devkit';
/** the LazyDriver implements vendor methods at construction but allows attaching NoVendorDriver later to be a full one*/
export class LazyDriver {
    constructor(signer) {
        this.signer = signer;
        this._driver = null;
    }
    get noVendor() {
        if (!this._driver) {
            throw new Error('thor driver is not ready');
        }
        return this._driver;
    }
    setNoVendor(driver) {
        this._driver = driver;
    }
    get genesis() {
        return this.noVendor.genesis;
    }
    get head() {
        return this.noVendor.head;
    }
    pollHead() {
        return this.noVendor.pollHead();
    }
    getBlock(revision) {
        return this.noVendor.getBlock(revision);
    }
    getTransaction(id, allowPending) {
        return this.noVendor.getTransaction(id, allowPending);
    }
    getReceipt(id) {
        return this.noVendor.getReceipt(id);
    }
    getAccount(addr, revision) {
        return this.noVendor.getAccount(addr, revision);
    }
    getCode(addr, revision) {
        return this.noVendor.getCode(addr, revision);
    }
    getStorage(addr, key, revision) {
        return this.noVendor.getStorage(addr, key, revision);
    }
    explain(arg, revision, cacheHints) {
        return this.noVendor.explain(arg, revision, cacheHints);
    }
    filterEventLogs(arg) {
        return this.noVendor.filterEventLogs(arg);
    }
    filterTransferLogs(arg) {
        return this.noVendor.filterTransferLogs(arg);
    }
    getFeesHistory(newestBlock, blockCount, rewardPercentiles) {
        return this.noVendor.getFeesHistory(newestBlock, blockCount, rewardPercentiles);
    }
    getPriorityFeeSuggestion() {
        return this.noVendor.getPriorityFeeSuggestion();
    }
    signTx(msg, options) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.signer.then(b => b.signTx(msg, options));
        });
    }
    signCert(msg, options) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.signer.then(b => b.signCert(msg, options));
        });
    }
}
const cache = {};
/**
 * create a no vendor driver
 * @param node the url of thor node
 * @param genesis the enforced genesis block
 */
export function createNoVendor(node, genesis) {
    const key = blake2b256(JSON.stringify({
        node,
        genesis
    })).toString('hex');
    let driver = cache[key];
    if (!driver) {
        cache[key] = driver = new DriverNoVendor(new SimpleNet(node), genesis);
    }
    return driver;
}
/**
 * create a full driver
 * @param node the url of thor node
 * @param genesis the enforced genesis block
 * @param newSigner a function to create signer
 */
export function createFull(node, genesis, newSigner) {
    const driver = new LazyDriver(newSigner(genesis.id));
    driver.setNoVendor(createNoVendor(node, genesis));
    return driver;
}
