// @vechain/connex/esm/index.js
import { Framework, newVendor } from '@vechain/connex-framework';
import { genesisBlocks } from './config';
import { createFull, createNoVendor, LazyDriver } from './driver';
import { createSync, createSync2 } from './signer';
/** convert options.network to Connex.Thor.Block */
function normalizeNetwork(n) {
    n = n || 'main';
    if (typeof n === 'string') {
        const gb = genesisBlocks[n];
        if (!gb) {
            throw new Error('invalid network');
        }
        return gb;
    }
    else {
        return n;
    }
}
/** convert network name to genesis id */
function normalizeGenesisId(id) {
    id = id || 'main';
    if (/^0x[0-9a-f]{64}$/.test(id)) {
        return id;
    }
    const gb = genesisBlocks[id];
    if (gb) {
        return gb.id;
    }
    throw new Error('invalid genesis id');
}
/** convert options.signer to a signer creation function */
function normalizeSigner(genesisId, signer) {
    switch (signer.toLocaleLowerCase()) {
        case 'sync':
            if (!window.connex) {
                throw new Error('Sync not found');
            }
            if (window.connex.thor.genesis.id !== genesisId) {
                throw new Error('Network mismatch');
            }
            return createSync;
        case 'sync2':
            return createSync2;
        default:
            throw new Error('unsupported signer');
    }
}
/** Thor class which can work stand alone to provide reading-services only */
class ThorClass {
    constructor(opts) {
        const genesis = normalizeNetwork(opts.network);
        const driver = createNoVendor(opts.node, genesis);
        const framework = new Framework(driver);
        return {
            get genesis() { return framework.thor.genesis; },
            get status() { return framework.thor.status; },
            get ticker() { return framework.thor.ticker.bind(framework.thor); },
            get account() { return framework.thor.account.bind(framework.thor); },
            get block() { return framework.thor.block.bind(framework.thor); },
            get transaction() { return framework.thor.transaction.bind(framework.thor); },
            get filter() { return framework.thor.filter.bind(framework.thor); },
            get explain() { return framework.thor.explain.bind(framework.thor); },
            get fees() {
                return {
                    get history() { return framework.thor.fees.history.bind(framework.thor.fees); },
                    get priorityFee() { return framework.thor.fees.priorityFee.bind(framework.thor.fees); },
                };
            },
        };
    }
}
/** Vendor class which can work standalone to provides signing-services only */
class VendorClass {
    constructor(genesisId, signer = 'sync2') {
        genesisId = normalizeGenesisId(genesisId);
        const newSigner = normalizeSigner(genesisId, signer);
        const driver = new LazyDriver(newSigner(genesisId));
        const vendor = newVendor(driver);
        return {
            get sign() {
                return vendor.sign.bind(vendor);
            }
        };
    }
}
/** Connex class */
class ConnexClass {
    constructor(opts) {
        var _a;
        const genesis = normalizeNetwork(opts.network);
        const newSigner = normalizeSigner(genesis.id, (_a = opts.signer) !== null && _a !== void 0 ? _a : 'sync2');
        const driver = createFull(opts.node, genesis, newSigner);
        const framework = new Framework(driver);
        return {
            get thor() { return framework.thor; },
            get vendor() { return framework.vendor; }
        };
    }
}
ConnexClass.Thor = ThorClass;
ConnexClass.Vendor = VendorClass;
export default ConnexClass;
export { ConnexClass as Connex };
