// @vechain/connex/esm/signer.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import { loadLibrary } from './script-loader';
import randomBytes from 'randombytes';
import { blake2b256 } from 'thor-devkit';
const BUDDY_SRC = 'https://unpkg.com/@vechain/connex-wallet-buddy@0.1';
const BUDDY_LIB_NAME = 'ConnexWalletBuddy';
export const createSync2 = (genesisId) => __awaiter(void 0, void 0, void 0, function* () {
    return loadLibrary(BUDDY_SRC, BUDDY_LIB_NAME).then(lib => lib.create(genesisId, () => randomBytes(16).toString('hex'), val => blake2b256(val).toString('hex')));
});
export const createSync = (genesisId) => __awaiter(void 0, void 0, void 0, function* () {
    const v1 = window.connex.vendor;
    return Promise.resolve({
        signTx: (msg, options) => {
            const s1 = v1.sign('tx');
            options.signer && s1.signer(options.signer);
            options.gas && s1.gas(options.gas);
            options.dependsOn && s1.dependsOn(options.dependsOn);
            options.link && s1.link(options.link);
            options.comment && s1.link(options.comment);
            if (options.delegator) {
                const url = options.delegator.url;
                s1.delegate((unsignedTx) => __awaiter(void 0, void 0, void 0, function* () {
                    const res = yield fetch(url, {
                        method: 'POST',
                        body: JSON.stringify(unsignedTx),
                        headers: {
                            "Content-Type": 'application/json'
                        }
                    });
                    return res.json();
                }));
            }
            options.onAccepted && options.onAccepted();
            return s1.request(msg);
        },
        signCert: (msg, options) => {
            const s1 = v1.sign('cert');
            options.signer && s1.signer(options.signer);
            options.link && s1.link(options.link);
            options.onAccepted && options.onAccepted();
            return s1.request(msg);
        }
    });
});
