// @vechain/connex-framework/dist/thor.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.newThor = void 0;
const V = require("validator-ts");
const account_visitor_1 = require("./account-visitor");
const block_visitor_1 = require("./block-visitor");
const explainer_1 = require("./explainer");
const fees_visitor_1 = require("./fees-visitor");
const filter_1 = require("./filter");
const head_tracker_1 = require("./head-tracker");
const R = require("./rules");
const tx_visitor_1 = require("./tx-visitor");
function newThor(driver) {
    const headTracker = head_tracker_1.newHeadTracker(driver);
    const readyDriver = (() => __awaiter(this, void 0, void 0, function* () {
        if (headTracker.head.number > 0) {
            return driver;
        }
        yield headTracker.ticker().next();
        return driver;
    }))();
    const genesis = JSON.parse(JSON.stringify(driver.genesis));
    return {
        get genesis() { return genesis; },
        get status() {
            return {
                head: headTracker.head,
                progress: headTracker.progress,
                finalized: headTracker.finalized
            };
        },
        ticker: () => headTracker.ticker(),
        account: addr => {
            addr = R.test(addr, R.address, 'arg0').toLowerCase();
            return account_visitor_1.newAccountVisitor(readyDriver, addr);
        },
        block: revision => {
            if (typeof revision === 'undefined') {
                revision = driver.head.id;
            }
            else {
                R.ensure(typeof revision === 'string' ? R.isHexBytes(revision, 32) : R.isUInt(revision, 32), 'arg0: expected bytes32 or unsigned 32-bit integer');
            }
            return block_visitor_1.newBlockVisitor(driver, typeof revision === 'string' ? revision.toLowerCase() : revision);
        },
        transaction: id => {
            id = R.test(id, R.bytes32, 'arg0').toLowerCase();
            return tx_visitor_1.newTxVisitor(readyDriver, id);
        },
        filter: (kind, criteria) => {
            R.ensure(kind === 'event' || kind === 'transfer', `arg0: expected 'event' or 'transfer'`);
            if (kind === 'event') {
                R.test(criteria, [eventCriteriaScheme], 'arg1');
                return filter_1.newFilter(readyDriver, 'event', criteria
                    .map(c => {
                    return {
                        address: c.address ? c.address.toLowerCase() : undefined,
                        topic0: c.topic0 ? c.topic0.toLowerCase() : undefined,
                        topic1: c.topic1 ? c.topic1.toLowerCase() : undefined,
                        topic2: c.topic2 ? c.topic2.toLowerCase() : undefined,
                        topic3: c.topic3 ? c.topic3.toLowerCase() : undefined,
                        topic4: c.topic4 ? c.topic4.toLowerCase() : undefined
                    };
                }));
            }
            else {
                R.test(criteria, [transferCriteriaScheme], 'arg1');
                return filter_1.newFilter(readyDriver, 'transfer', criteria
                    .map(c => {
                    return {
                        txOrigin: c.txOrigin ? c.txOrigin.toLowerCase() : undefined,
                        sender: c.sender ? c.sender.toLowerCase() : undefined,
                        recipient: c.recipient ? c.recipient.toLowerCase() : undefined
                    };
                }));
            }
        },
        explain: (clauses) => {
            R.test(clauses, [clauseScheme], 'arg0');
            return explainer_1.newExplainer(readyDriver, clauses);
        },
        fees: {
            history: (newestBlock) => {
                if (typeof newestBlock === 'undefined') {
                    newestBlock = driver.head.id;
                }
                else {
                    R.ensure(typeof newestBlock === 'string' ? R.isHexBytes(newestBlock, 32) : R.isUInt(newestBlock, 32), 'arg0: expected bytes32 or unsigned 32-bit integer');
                }
                return fees_visitor_1.newFeesHistoryVisitor(driver, newestBlock);
            },
            priorityFee: () => driver.getPriorityFeeSuggestion()
        }
    };
}
exports.newThor = newThor;
const clauseScheme = {
    to: V.nullable(R.address),
    value: R.bigInt,
    data: V.optional(R.bytes)
};
const eventCriteriaScheme = {
    address: V.optional(R.address),
    topic0: V.optional(R.bytes32),
    topic1: V.optional(R.bytes32),
    topic2: V.optional(R.bytes32),
    topic3: V.optional(R.bytes32),
    topic4: V.optional(R.bytes32)
};
const transferCriteriaScheme = {
    sender: V.optional(R.address),
    recipient: V.optional(R.address),
    txOrigin: V.optional(R.address)
};
